File: src/main/java/edu/kit/ifv/mobitopp/actitopp/ActivityType.java
package edu.kit.ifv.mobitopp.actitopp;

import java.util.Collections;
import java.util.EnumSet;
import java.util.Set;

public enum ActivityType {
	
	WORK ('W'),
	EDUCATION ('E'),
	LEISURE ('L'),
	SHOPPING ('S'),
	TRANSPORT ('T'),
	HOME ('H'),
	UNKNOWN('x')
	;
	

	public static final Set<ActivityType> OUTOFHOMEACTIVITY;
	public static final Set<ActivityType> FULLSET;
	
	static 
	{
		OUTOFHOMEACTIVITY = Collections.unmodifiableSet(EnumSet.of(
																														ActivityType.WORK,
																														ActivityType.EDUCATION,
																														ActivityType.LEISURE,
																														ActivityType.SHOPPING,
																														ActivityType.TRANSPORT
																									));
		
		FULLSET = Collections.unmodifiableSet(EnumSet.of(
																														ActivityType.WORK,
																														ActivityType.EDUCATION,
																														ActivityType.LEISURE,
																														ActivityType.SHOPPING,
																														ActivityType.TRANSPORT,
																														ActivityType.HOME
																										));
	}
	
	
	private char charValue;
	
	private ActivityType(char charValue) 
	{
		this.charValue=charValue;	
	}
	
	public char getTypeasChar() 
	{ 
		return this.charValue; 
	}

	public static ActivityType getTypeFromChar(char charValue) 
	{
		Character tocompare = Character.toUpperCase(charValue);
		
		for (ActivityType type : EnumSet.allOf(ActivityType.class)) 
		{
			if (type.getTypeasChar() ==  tocompare) 
			{
				return type;
			}
		}
		return null;
	}
}

File: src/main/java/edu/kit/ifv/mobitopp/actitopp/enums/ActivityType.kt
package edu.kit.ifv.mobitopp.actitopp.enums
import java.util.Collections
import java.util.EnumSet
// TODO find out whether Unknown is deliberately set to a lower case letter so that the getTypeFromChar Throws
enum class ActivityType(val typeasChar: Char, val defaultActivityTime: Int = 278) {
    EDUCATION('E', 340),
    HOME('H'),
    LEISURE('L', 130),
    SHOPPING('S', 41),
    TRANSPORT('T', 15),
    WORK('W', 472),
    UNKNOWN('x');


    companion object {
        val OUTOFHOMEACTIVITY: Set<ActivityType> = Collections.unmodifiableSet(
            EnumSet.of(
                WORK,
                EDUCATION,
                LEISURE,
                SHOPPING,
                TRANSPORT
            )
        )

        val FULLSET: Set<ActivityType> = Collections.unmodifiableSet(
            EnumSet.of(
                EDUCATION,
                HOME,
                LEISURE,
                SHOPPING,
                TRANSPORT,
                WORK,
            )
        )

        fun getTypeFromChar(charValue: Char): ActivityType {
            return entries.firstOrNull {it.typeasChar == charValue.uppercaseChar()} ?: throw NoSuchElementException("There is no type $charValue as activity in actiTopp")

        }
    }
}
